sources/packages/backend/apps/queue-consumers/src/processors/schedulers/base-scheduler.ts

```diff
@@ -84,29 +87,81 @@ export abstract class BaseScheduler<T> implements OnApplicationBootstrap {
+    // queue-consumers instances to be initialized at the same time
+    // and execute the checks concurrently.
+    await this.queueService.acquireQueueLock(
+      this.schedulerQueue.name as QueueNames,
+      async () => {
+        await this.ensureNextDelayedJobCreation();
+      },
+    );
   }
 
   /**
-   * Check if there is any old cron job  (i.e whenever there is a
-   * change in cron option, then a new job is created the old job
-   * will be still there in the queue) and delete it and add the
-   * new job to the queue.
-   * Note: If there is an old retrying job, it won't be deleted,
-   * as "getRepeatableJobs" will not fetch retrying jobs.
+   * Ensures a scheduled job will have a delayed job created with the
+   * next expected scheduled time based on the configured cron expression.
    */
-  private async deleteOldRepeatableJobs(): Promise<void> {
-    const getAllRepeatableJobs = await this.schedulerQueue.getRepeatableJobs();
-    const cronRepeatOption = await this.queueCronConfiguration();
-    getAllRepeatableJobs.forEach((job) => {
-      if (job.cron !== cronRepeatOption.cron) {
-        this.schedulerQueue.removeRepeatableByKey(job.key);
+  private async ensureNextDelayedJobCreation(): Promise<void> {
+    const delayedJobs = await this.schedulerQueue.getDelayed();
+    const nextExpectedScheduledJobDate =
+      await this.getNexSchedulerExecutionDate();
+    // Check if there is a delayed job with the expected scheduled time.
+    const expectedDelayedJob = delayedJobs.find((delayedJob) => {
+      const nextExecution = this.getDelayedJobExecutionDate(delayedJob);
+      return dayjs(nextExecution).isSame(nextExpectedScheduledJobDate);
+    });
+    // If the only delayed job is the expected one, no further verifications are needed.
+    if (expectedDelayedJob && delayedJobs.length === 1) {
+      return;
+    }
+    // Remove any non expected delayed job.
+    for (const delayedJob of delayedJobs) {
+      if (!expectedDelayedJob || delayedJob !== expectedDelayedJob) {
+        await delayedJob.remove();
       }
+    }
+    // The expected delayed job was found and any
+    // extra delayed jobs were removed.
+    if (expectedDelayedJob) {
+      return;
+    }
+    // Creating a unique job ID ensures that the delayed jobs
+    // will be created even if they were already promoted.
+    const uniqueJobId = `${this.schedulerQueue.name}:${uuid()}`;
+    await this.schedulerQueue.add(await this.payload(), {
+      jobId: uniqueJobId,
     });
   }
 
+  /**
+   * Gets the next scheduled date for the given cron expression.
+   * @param cron cron expression string.
+   * @returns the next date the scheduler will be executed.
+   */
+  async getNexSchedulerExecutionDate(): Promise<Date> {
+    const repeatOptions = await this.queueCronConfiguration();
+    const result = cronParser.parseExpression(repeatOptions.cron, {
+      utc: true,
+    });
+    return result.next().toDate();
```

Comment:
We can return milliseconds here `result.next().getMilliseconds()` and do a millisecond comparison between `nextSchedulerExecutionDate` and `delayedJobExecutionDate`

Reply:
It would work but IMO it would be clear to have a proper date.
Is the idea to avoid using the dayjs?
I will give it a try to see if the readability is still good enough.

Reply:
The `getMilliseconds` does not contain the total milliseconds.
I would prefer to keep the current implementation unless it would be a bloker.

Reply:
Tested with `getDate` (similar to the Date) and it seems all good.